### DISCUSSION TITLE
Content Pipeline and dependent files

### DISCUSSION THREAD

**Discussion by besopuzzz** [MonoGame/MonoGame#8168] [Content]
Can you tell me how to implement the construction of individual files? Let me give you an example: there is an atlas in which individual sprite regions are stored. When the atlas file is being built, separate sprites must be created and rebuilt (with a specific Importer and Proccessor). I looked at the ContentProcessor methods and found AddOutputFile() and addDependency() there. I think they will help me, but I don't even have a rough understanding of how these methods work.

**Comment by ThomasFOG** [comment]
I am not sure about what you are trying to do, but following some of your messages it seems that you are trying to work with texture atlases, though not a way you're intended to work with.

Tell me if I understand it wrongly, but I feel like you are trying to give a spritesheet (a single image file containing multiple frames of sprites) to the MGCB, and wish that to be treated as if each of those frames can be loaded individually like ```Content.Load()``` would only load a single of those frame, am I correct?

This sounds like a misunderstanding of how spritesheet work and why they exist.

Spritesheets are expected to stay a single unique texture/file with a single ```Content.Load()```, and then you would draw the appropriate frame by using a source rectangle with the proper ```SpriteBatch.Draw()``` overload.
This is more optimized because asking the GPU to change texture might be a costly operation impairing performances.

What you are likely looking into, is how to define those source rectangle and retrieve them at runtime, independently of how the texture is treated. A possible solution would be to simply put the rectangles position and size in a text file, and load that text file like a standard file and parse those rectangles. Some software export this for you, like Aseprite, which can export a spritesheet with a PNG containing all the frames, and a JSON file containing the rectangles. You would them load the PNG through MGCB, and the JSON manually. You could also look into [MonoGame.Aseprite](https://monogameaseprite.net/) which propose importers and processors so that all of that is transparent to you (but before using it, I would suggest  try doing it the manual way to assimilate the concept).

If I'm wrong and you are really looking into splitting a spritesheet into individual files, then MGCB is likely not what you are looking for and I would rather look into how to split PNGs.

**Reply by besopuzzz** [reply]
I understand perfectly how SpriteSheet works. I don't want to cut the PNG file into small sprites. Ideally, I wanted to analyze the atlas into sprites, serialize them into files and then load them in the game via ContentManager.Load(). Moreover, each Sprite would have a link to the atlas (and already in it access to a common texture). 

Although if we talk about the initial problem, I wanted to load atlas sprites via ContentManager.Load() via the path "Atlases/AtlasLevel1_40", where Atlases/ is the folder with atlases, AtlasLevel1 is the atlas of level 1, _40 is the sprite index. But since ContentManager.Load() first checks for the presence of a file along the path "Atlases/AtlasLevel1_40", I will get an error. From this, I got the idea to create these sprites manually via MGCB. But I understand that MGCB is too stripped down and does not have the functionality to expand and use user parameters (besides .dll files), and I will not be able to solve my problem this way.

**Reply by ThomasFOG** [reply]
I see. MGCB is merely a format converter and serializer. You could kind of bend it to your use case with some hacks, but I feel like the solution would be more convoluted than it has to be.

I would load AtlasLevel1 as a normal texture through MGCB, and then use System.IO to load and parse the _40 file.
